UEFI

N00byEdge opened the pull request.

N00byEdge committed:
Remove platform.ElfType

N00byEdge committed:
WIP UEFI support

N00byEdge committed:
ELF loading: Always pass in paging root

N00byEdge committed:
UEFI: Export pub panic

N00byEdge committed:
Paging: fix formatting

N00byEdge committed:
Initial UEFI implementation is done

N00byEdge committed:
Move uefi.sh into build.zig

N00byEdge committed:
PMM: Make UEFI check comptime

StaticSaga commented on src/platform/uefi_aarch64/main.zig:
```diff
@@ -0,0 +1,369 @@
+pub fn handleProtocol(handle: uefi.Handle, comptime T: type) callconv(.Inline) ?*T {
+    var ptr: *T = undefined;
+    const guid: std.os.uefi.Guid align(8) = T.guid;
+    if (.Success != uefi.system_table.boot_services.?.handleProtocol(handle, &guid, @ptrCast(*?*c_void, &ptr))) {
```
Why is the order inverted here?

N00byEdge replied:
Because where I stole it from had it in that order :^)

StaticSaga commented on src/platform/uefi_aarch64/main.zig:
```diff
@@ -0,0 +1,369 @@
+    sabaton.paging.apply_paging(&paging_root);
+
+    asm volatile (
+        \\  MSR SPSel, XZR
```
Maybe refactor this into a separate function

N00byEdge replied:
Fair

StaticSaga commented on uefidir/image/startup.nsh:
```diff
@@ -0,0 +1 @@
+FS0:\EFI\BOOT\BOOTA64.EFI
```
Why is startup.nsh necessary here? It should autoboot EFI\BOOT\BOOTA64.EFI.

N00byEdge replied:
It doesn't autoboot it for me though, and I'm not sure why. Are my efi vars just wrong?

StaticSaga replied:
Very likely, but I am not sure how or why. I had the same issue with x86 uefi and it somehow resolved with a totally unrelated change

N00byEdge replied:
Can confirm, fixing my boot settings makes it work.

StaticSaga reviewed (commented):
Looks good to me

N00byEdge committed:
More consistent equality check operand order

N00byEdge committed:
Break entering kernel into its own function

StaticSaga reviewed (commented):
Maybe change the build mode to ReleaseSmall to be consistent.

N00byEdge committed:
UEFI: Remove startup.nsh

N00byEdge committed:
Make sortStivale2Memmap iterative

The tailcall was cool but it bugged out on Release{Fast,Small}, thanks
for another zig compiler bug.

N00byEdge committed:
UEFI: Build with .ReleaseSmall

N00byEdge committed:
UEFI: Add usb input to qemu for configuration editing
